sycl/doc/extensions/Matrix/dpcpp-joint-matrix.asciidoc

```diff
@@ -0,0 +1,245 @@
+|1     |Initial extension implementation on AMX.  Base features are supported.
+|======================
+
+## New `matrix` class
+We introduce a new class called `matrix`. The user needs to specify the type of the elements, sizes, and the memory layout.`matrix` is distributed among an execution unit. In practice this can be one work-item, the work-items in a sub-group, or the work-items in a work-group.
+
+	
+The same class `matrix` should handle both cases where sizes are constant (GPU case) and when sizes are variables (CPU case). Note that a AMX 2d tile register permits sizes up to 1024 (16rowsx64cols) bytes. The ability to define only one interface for both makes it possible to give the user a way to make use of the flexibility introduced by the CPU but at the same time save resources on the GPU. We use `sycl::dynamic_extent`  to differentiate between static and dynamic sizes.
+	
+Layout is necessary on the type to be able to calculate the physical offset if the user needs to access a single entry for some purpose. Besides row major and column major layout, `matrix_layout` is flexible enough to introduce customed layouts such as symmetric or tiled layouts. AMX hardware requires A and B to be in VNNI or 32 bits packed layout. If users did not specify these layouts, transforms done by the implementation will be slow due to extra scatter/gather operations. Hence, we expose these layouts `packed_a` and `packed_b` to the user to specify that A and/or B have already been VNNIed. The packed or VNNI layout is introduced in `VNNI layout` section below.
+	
+```c++
+namespace sycl::ext::intel::experimental::matrix {
+template <typename T, size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent, matrix_layout l = row_major>
+struct matrix;
+enum class matrix_layout {
+  row_major,
+  col_major,
+  packed_a,
+  packed_b
+};
+}
+```
+
+
+### Memory Scope
+For the memory scope, we have two solutions. The long term solution is to use the proposed https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/LocalMemory/SYCL_INTEL_local_memory.asciidoc[`group_local_memory` extension] to allocate the matrix in local memory associated with a SYCL group as shown in the example below.
+
+```c++
+multi_ptr<matrix<T>, address_space::local_space> tA_ptr = group_local_memory<matrix<sub_group, int8_t, tM, tN>>(sg);
+```
+However, sub-group local memory is not yet well defined in {dpcpp}. Moreover, the representation of this notion in LLVM IR and SPIR-V is not yet clear. Hence, for this proposal, we use `joint_matrix` instead of `matrix` to emphasis that the matrix is shared among a group of work items and is not private to each work item. Moreover, the memory scope is added as an additional constructor argument. This results into the following description:
+
+```c++
+namespace sycl::ext::intel::experimental::matrix {
+template <typename Group, typename T, size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent, matrix_layout l = row_major>
```

Comment:
The API allows dynamic extents but the implementation may not. Note that dynamic sizes are supported by AMX. This will be part of the query. Remember the original "get_info"-based query design that we had first, had the following type:
 struct matrix_properties {
bool dynamic_p;
uint32_t msize; // row major
uint32_t  nsize; // row major
uint32_t  ksize; // row major
matrix_type atype; 
matrix_type btype; 
matrix_type ctype;  
memory_scope scope; 
};
dynamic_p was supposed to tell the user whether dynamic sizes are supported by the implementation or not.

While we dropped the "get_info"-based query because it cannot return constants, the current design using "engine query" could provide the same functionality. Please see the https://gitlab.devtools.intel.com/SYCL/extensions/-/issues/106 option2.

Reply:
I thought the primary goal of this document is to describe the API as it exists today, so that customers can use it in their code.  If that is the case, we shouldn't document features that don't exist.  Instead, list these future extensions below under "Future-looking API".

In this particular case, it will be very easy to add the default argument of `sycl::dynamic_extent` later.  This won't break any existing code.  By contrast, leaving `sycl::dynamic_extent` as the default parameter now just leads customers to believe the feature exists when that is not the case.

Reply:
The API as it exists today has the dynamic extent syntax. It has been there for while to be as close to the C++ syntax as possible. But the implementation support is restrictive to only static sizes. We said we only support what the hardware support + some implementation defined choices. The query will help clear these out for the user once we add it. In the meantime, we document what the implementation supports in the section: Current Implementation Restrictions. I noticed I don't say that explicitly in this section. I will add it explicitly in that section, that the current implementation only supports static sizes.